### PR TITLE
[MC/DC] Refactor: Make `MCDCParams` as `std::variant`

### PR DESCRIPTION
**Pull request by chapuni**
Introduce `mcdc::DecisionParameters` and `mcdc::BranchParameters` and make sure them not initialized as zero.

FIXME: Could we make `CoverageMappingRegion` as a smart tagged union?

### CODE REVIEW

File: llvm/lib/ProfileData/Coverage/CoverageMapping.cpp [hunk 1]
```diff
@@ -264,9 +266,10 @@ class MCDCRecordProcessor {
...
                       ArrayRef<const CounterMappingRegion *> Branches)
-      : Bitmap(Bitmap), Region(Region), Branches(Branches),
-        NumConditions(Region.MCDCParams.NumConditions),
-        BitmapIdx(Region.MCDCParams.BitmapIdx * CHAR_BIT),
+      : Bitmap(Bitmap), Region(Region),
+        DecisionParams(Region.getDecisionParams()), Branches(Branches),
+        NumConditions(DecisionParams.NumConditions),
+        BitmapIdx(DecisionParams.BitmapIdx * CHAR_BIT),
         Folded(NumConditions, false), IndependencePairs(NumConditions),
         TestVectors((size_t)1 << NumConditions) {}
```
Comment (chapuni) [inline llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:274]: I'll make this mergeable later, possibly tonight.

File: llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h [hunk 2]
```diff
@@ -250,18 +251,27 @@ struct CounterMappingRegion {
...
     /// IDs used to represent a branch region and other branch regions
     /// evaluated based on True and False branches.
-    MCDCConditionID ID = 0, TrueID = 0, FalseID = 0;
+    MCDCConditionID ID, TrueID, FalseID;
+
+    MCDCBranchParameters() = delete;
   };
 
+  using MCDCParameters = std::variant<std::monostate, MCDCDecisionParameters,
+                                      MCDCBranchParameters>;
```
Comment (chapuni) [inline llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h:273]: FYI, the size of `MCDCParameters` will be `12` (gnu) or `8` (libc++) finally, with shrinking a few members to `int16_t`. (#81257)

File: llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp [hunk 3]
```diff
@@ -308,13 +309,21 @@ Error RawCoverageMappingReader::readMappingRegionsSubArray(
             return Err;
           if (auto Err = readIntMax(FID, std::numeric_limits<unsigned>::max()))
             return Err;
+          if (ID == 0)
+            return make_error<CoverageMapError>(
+                coveragemap_error::malformed,
+                "MCDCConditionID shouldn't be zero");
+          Params = CounterMappingRegion::MCDCBranchParameters{
+              (unsigned)ID, (unsigned)TID, (unsigned)FID};
```
Comment (ornata) [inline llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:317]: `static_cast`?

**Comment by ornata** [comment]
Is the main benefit of this avoiding zero initialization?

File: llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp (hunk 3 above)
Comment (chapuni) [reply llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:317]: I was so lazy since their casts would be removed by other PRs. I'll do shortly.

**Review by chapuni — COMMENTED** [review]
@ornata I want to isolate Decision stuff and Branch stuff in `MCDCParams`.
Also I'd like to encapsulate params into each record. Once parameters are set, they are expected to hold "valid" values. Zeroing really confused me, (possibly us).

I've found `std::variant` works good for this area. I expect it will mitigate confusions little a bit.

**Review by ornata — APPROVED** [review]
LGTM

**Comment by LW-archlinux** [comment]
Commit 75f0d40507ea3f7c99dd3250ff0fbe6dab341910 breaks build for a multilib 32-bit clang build.

See https://github.com/llvm/llvm-project/issues/81880

**Pull request merged**